ambry-router/src/main/java/com.github.ambry.router/GetOperation.java

```diff
@@ -0,0 +1,214 @@
+   *                                    created as part of this poll operation.
+   */
+  abstract void poll(RequestRegistrationCallback<GetOperation> requestRegistrationCallback);
+
+  /**
+   * Handle the given {@link ResponseInfo} received for a request that was sent out.
+   * @param responseInfo the {@link ResponseInfo} to be handled.
+   */
+  abstract void handleResponse(ResponseInfo responseInfo);
+
+  /**
+   * Process the given {@link ServerErrorCode} and set operation status accordingly.
+   * @param errorCode the {@link ServerErrorCode} to process.
+   */
+  void processServerError(ServerErrorCode errorCode) {
+    switch (errorCode) {
+      case Blob_Deleted:
+        logger.trace("Requested blob was deleted");
+        setOperationException(new RouterException("Server returned: " + errorCode, RouterErrorCode.BlobDeleted));
+        break;
+      case Blob_Expired:
+        logger.trace("Requested blob is expired");
+        setOperationException(new RouterException("Server returned: " + errorCode, RouterErrorCode.BlobExpired));
+        break;
+      case Blob_Not_Found:
+        logger.trace("Requested blob was not found on this server");
+        setOperationException(new RouterException("Server returned: " + errorCode, RouterErrorCode.BlobDoesNotExist));
+        break;
+      default:
+        setOperationException(
+            new RouterException("Server returned: " + errorCode, RouterErrorCode.UnexpectedInternalError));
+        break;
+    }
+  }
+
+  /**
+   * Set the exception associated with this operation.
+   * A {@link ServerErrorCode#Blob_Deleted} or {@link ServerErrorCode#Blob_Expired} error overrides any other
+   * previously received exception.
+   * @param exception the {@link RouterException} to possibly set.
+   */
+  void setOperationException(RouterException exception) {
+    if (operationException == null || exception.getErrorCode() == RouterErrorCode.BlobDeleted
+        || exception.getErrorCode() == RouterErrorCode.BlobExpired) {
+      operationException = exception;
+    }
+  }
+
+  /**
+   * Create and return the {@link GetRequest} associated with the given blobId.
+   * @return the created {@link GetRequest}.
+   * @param blobId The {@link BlobId} for which the {@link GetRequest} is being created.
+   * @param flag The {@link MessageFormatFlags} to be set with the GetRequest.
+   * @return the created GetRequest.
+   */
+  protected GetRequest createGetRequest(BlobId blobId, MessageFormatFlags flag) {
+    List<BlobId> blobIds = Collections.singletonList(blobId);
+    List<PartitionRequestInfo> partitionRequestInfoList = Collections.singletonList(new PartitionRequestInfo(blobId
+        .getPartition(), blobIds));
+    return new GetRequest(NonBlockingRouter.correlationIdGenerator.incrementAndGet(), routerConfig.routerHostname, flag,
+        partitionRequestInfoList, GetOptions.None);
+  }
+}
+
+/**
+ * A class that holds information about the get requests sent out.
+ */
+class GetRequestInfo {
```

Comment:
These are just simple struct variables really. The alternative is to make this a full fledged class outside of all these classes, which I think is unnecessary at this time.